nimbus/src/enrollment.rs

```diff
@@ -311,55 +758,820 @@ mod tests {
+        let mut events = vec![];
+        let enrollment_id = Uuid::new_v4();
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug.clone(),
+            status: EnrollmentStatus::Enrolled {
+                enrollment_id,
+                branch: "control".to_owned(),
+                reason: EnrolledReason::Qualified,
+            },
+        };
+        let enrollment = evolver
+            .evolve_enrollment(
+                true,
+                Some(&exp),
+                Some(&exp),
+                Some(&existing_enrollment),
+                &mut events,
+            )?
+            .unwrap();
+        if let EnrollmentStatus::Disqualified {
+            reason: DisqualifiedReason::NotTargeted,
+            enrollment_id: new_enrollment_id,
+            ..
+        } = enrollment.status
+        {
+            assert_eq!(new_enrollment_id, enrollment_id);
+        } else {
+            panic!("Wrong variant!");
+        }
+        assert_eq!(events.len(), 1);
+        assert_eq!(events[0].enrollment_id, enrollment_id.to_string());
+        assert_eq!(events[0].experiment_slug, exp.slug);
+        assert_eq!(events[0].branch_slug, "control");
+        assert_eq!(events[0].reason, Some("targeting".to_owned()));
+        assert_eq!(
+            events[0].change,
+            EnrollmentChangeEventType::Disqualification
+        );
+        Ok(())
     }
 
-    fn get_experiment_enrollments_with_status(
-        db: &Database,
-        status: EnrollmentStatus,
-    ) -> Result<Vec<ExperimentEnrollment>> {
-        let enrollments = get_experiment_enrollments(db)?;
-        Ok(enrollments
-            .into_iter()
-            .filter(|enr| enr.status == status)
-            .collect())
+    #[test]
+    fn test_evolver_experiment_update_enrolled_then_branches_changed() -> Result<()> {
+        let mut exp = get_test_experiments()[0].clone();
+        exp.branches = vec![crate::Branch {
+            slug: "bobo-branch".to_owned(),
+            ratio: 1,
+            feature: None,
+        }];
+        let (nimbus_id, app_ctx, aru) = local_ctx();
+        let evolver = enrollment_evolver(&nimbus_id, &app_ctx, &aru);
+        let mut events = vec![];
+        let enrollment_id = Uuid::new_v4();
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug.clone(),
+            status: EnrollmentStatus::Enrolled {
+                enrollment_id,
+                branch: "control".to_owned(),
+                reason: EnrolledReason::Qualified,
+            },
+        };
+        let enrollment = evolver
+            .evolve_enrollment(
+                true,
+                Some(&exp),
+                Some(&exp),
+                Some(&existing_enrollment),
+                &mut events,
+            )?
+            .unwrap();
+        if let EnrollmentStatus::Disqualified {
+            reason: DisqualifiedReason::NotTargeted,
```

Comment:
As noted in a comment above, I think the expected behaviour here is that we stay in the same branch rather than becoming disqualified.